pandas/io/formats/style.py

```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
+        tooltips to render in HTML.
+
+        Returns
+        -------
+        styles : List
+        """
+        return [{"selector": f".{self.class_name}", "props": self.class_properties}]
+
+    def _translate_tooltips(self, styler_data, uuid, d):
+        """
+        Mutate the render dictionary to allow for tooltips:
+
+        - Add `<span>` HTML element to each data cells `display_value`. Ignores
+          headers.
+        - Add table level CSS styles to control pseudo classes.
+
+        Parameters
+        ----------
+        styler_data : DataFrame
+            Underlying ``Styler`` DataFrame used for reindexing.
+        uuid : str
+            The underlying ``Styler`` uuid for CSS id.
+        d : dict
+            The dictionary prior to final render
+        """
+        self.tt_data = (
+            self.tt_data.reindex_like(styler_data)
+            .dropna(how="all", axis=0)
+            .dropna(how="all", axis=1)
+        )
+        if self.tt_data.empty:
+            return d
+
+        name = self.class_name
+
+        def _pseudo_css(row: int, col: int, text: str):
+            """
+            For every table data-cell that has a valid tooltip (not None, NaN or
+            empty string) must create two pseudo CSS entries for the specific
+            <td> element id which are added to overall table styles:
+            an on hover visibility change and a content change
+            dependent upon the user's chosen display string.
+
+            For example:
+                [{"selector": "T__row1_col1:hover .pd-t",
+                 "props": [("visibility", "visible")]},
+                {"selector": "T__row1_col1 .pd-t::after",
+                 "props": [("content", "Some Valid Text String")]}]
+
+            Parameters
+            ----------
+            row : int
+                The row index of the specified tooltip string data
+            col : int
+                The col index of the specified tooltip string data
+            text : str
+                The textual content of the tooltip to be displayed in HTML.
+
+            Returns
+            -------
+            pseudo_css : List
+            """
+            return [
+                {
+                    "selector": "#T_"
+                    + uuid
+                    + "row"
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f":hover .{name}",
+                    "props": [("visibility", "visible")],
+                },
+                {
+                    "selector": "#T_"
+                    + uuid
+                    + "row"
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f" .{name}::after",
+                    "props": [("content", f'"{text}"')],
+                },
+            ]
+
+        mask = (self.tt_data.isna()) | (self.tt_data.eq(""))  # empty string = no ttip
+        self.table_styles = [
```

Comment:
Can you just break this into separate for loops / functions? It's a little hard to grok what is going on here with the deeply nested comprehensions

Reply:
The objective is for `Styler.table_styles` to be extended by `_Tooltip.table_styles`.
But each data cell (if has tooltip) needs two table styles, so the method `_pseudo_css` generates the list of 2 items for each data cell.
The nesting is only needed to flatten the list to allow the above objective in the right format: `[[a1,b1],[a2,b2]] -> [a1, b1, a2, b2]`.

Happy to take suggestions on structure, but previous versions have had requests for list comprehensions as opposed to separate loops.